File: FitControl/mailservice/mailservice/src/main/java/com/fitcontrol/mailservice/controller/MailController.java
package com.fitcontrol.mailservice.controller;

import com.fitcontrol.mailservice.service.MailService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.Map;
import java.util.List;
import java.util.ArrayList;

@RestController
@RequestMapping("/api/mail")
public class MailController {

    @Autowired
    private MailService mailService;

    @PostMapping("/import-csv")
    public ResponseEntity<String> importCsv(
            @RequestParam("file") MultipartFile file,
            @RequestParam("subject") String subject,
            @RequestParam("body") String body
    ) {
        try {
            mailService.sendEmailsFromCsv(file.getInputStream(), subject, body);
            return ResponseEntity.ok("Correos enviados correctamente");
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("Error: " + e.getMessage());
        }
    }

    @PostMapping("/send-single")
    public ResponseEntity<String> sendSingle(@RequestBody Map<String, String> request) {
        try {
            String email   = request.get("email");
            String nombre  = request.get("nombre");
            String subject = request.get("subject");
            String body    = request.get("body");

            mailService.sendSingleEmail(email, nombre, subject, body);
            return ResponseEntity.ok("Correo enviado correctamente");
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("Error: " + e.getMessage());
        }
    }

    @PostMapping("/send-multiple")
    public ResponseEntity<String> sendMultiple(@RequestBody Map<String, Object> request) {
        try {
            String subject = (String) request.get("subject");
            String body = (String) request.get("body");
            List<Map<String, String>> recipientsList = (List<Map<String, String>>) request.get("recipients");

            if (recipientsList == null || recipientsList.isEmpty()) {
                return ResponseEntity.badRequest().body("No se proporcionaron destinatarios");
            }

            int successCount = 0;
            int failCount = 0;
            List<String> errors = new ArrayList<>();

            for (Map<String, String> recipient : recipientsList) {
                try {
                    String email = recipient.get("email");
                    String nombre = recipient.get("nombre");
                    mailService.sendSingleEmail(email, nombre, subject, body);
                    successCount++;
                } catch (Exception e) {
                    failCount++;
                    errors.add(recipient.get("email") + ": " + e.getMessage());
                }
            }

            String message = String.format("Enviados: %d | Fallidos: %d", successCount, failCount);
            if (!errors.isEmpty()) {
                message += " | Errores: " + String.join("; ", errors);
            }

            return ResponseEntity.ok(message);
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("Error general: " + e.getMessage());
        }
    }
}

File: FitControl/FitControl/app/Http/Controllers/MailController.php
<?php

namespace App\Http\Controllers;

use App\Mail\FitControlMail;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Mail;

class MailController extends Controller
{
    /**
     * Import CSV and send emails to all recipients.
     */
    public function importCsv(Request $request)
    {
        $request->validate([
            'file'    => 'required|file|mimes:csv,txt',
            'subject' => 'required|string',
            'body'    => 'required|string',
        ]);

        $subject = $request->input('subject');
        $body    = $request->input('body');
        
        // Guardar el archivo temporalmente
        $path = $request->file('file')->store('imports/csv', 'local');

        // Despachar el Job en segundo plano
        \App\Jobs\ProcessCsvImportJob::dispatch($path, $subject, $body);

        return response('El archivo CSV se está procesando en segundo plano.', 202);
    }

    /**
     * Send a single email.
     */
    public function sendSingle(Request $request)
    {
        $request->validate([
            'email'   => 'required|email',
            'nombre'  => 'required|string',
            'subject' => 'required|string',
            'body'    => 'required|string',
        ]);

        $email   = $request->input('email');
        $nombre  = $request->input('nombre');
        $subject = $request->input('subject');
        $body    = $request->input('body');

        try {
            \App\Jobs\SendBulkEmail::dispatch($email, $nombre, $subject, $body);

            return response('Correo enviado correctamente (encolado)');
        } catch (\Exception $e) {
            return response("Error: {$e->getMessage()}", 500);
        }
    }

    /**
     * Send batch emails to multiple recipients.
     */
    public function sendMultiple(Request $request)
    {
        $request->validate([
            'subject'   => 'required|string',
            'body'      => 'required|string',
            'recipients' => 'required|array|min:1',
            'recipients.*.email'  => 'required|email',
            'recipients.*.nombre' => 'required|string',
        ]);

        $subject = $request->input('subject');
        $body    = $request->input('body');
        $recipientsList = $request->input('recipients');

        $successCount = 0;
        $failCount = 0;
        $errors = [];

        foreach ($recipientsList as $recipient) {
            try {
                $email  = $recipient['email'];
                $nombre = $recipient['nombre'];
                \App\Jobs\SendBulkEmail::dispatch($email, $nombre, $subject, $body);
                $successCount++;
            } catch (\Exception $e) {
                $failCount++;
                $errors[] = "{$recipient['email']}: {$e->getMessage()}";
            }
        }

        $message = "Enviados: {$successCount} | Fallidos: {$failCount}";
        if (!empty($errors)) {
            $message .= " | Errores: " . implode('; ', $errors);
        }

        return response($message);
    }
}
